libpod/image/image.go

```diff
@@ -1075,3 +1077,176 @@ func (i *Image) Comment(ctx context.Context, manifestType string) (string, error
+		buffer.WriteString(fmt.Sprintf("Size         : %v\n", units.HumanSizeWithPrecision(float64(imageInfo.Size), 4)))
+		buffer.WriteString(fmt.Sprintf("Image Layers :\n"))
+		buffer.WriteString(fmt.Sprintf("%s", treeData))
+
+	}
+	return &buffer, nil
+}
+
+// Builds a Map of layerID to LayerInfo ( adds links Parent layer and child layers for each layer.)
+func buildImageLayersDependencyMap(ctx context.Context, i *Image, whatrequires bool) (map[string]*LayerInfo, error) {
+	imgs, _ := i.imageruntime.store.Images()
+	layers, _ := i.imageruntime.store.Layers()
+
+	// Builds the layer dependency chain with ParentID and ChildID
+	layerInfoMap := make(map[string]*LayerInfo)
+	for _, layer := range layers {
+		existingLayer, ok := layerInfoMap[layer.ID]
+		if !ok {
+			layerInfoMap[layer.ID] = &LayerInfo{
+				ID:       layer.ID,
+				Size:     layer.UncompressedSize,
+				ParentID: layer.Parent,
+			}
+
+			//Add current Layer as child of Parent layer.
+			existingLayer, ok = layerInfoMap[layer.Parent]
+			if !ok {
+				layerInfoMap[layer.Parent] = &LayerInfo{
+					ID: layer.Parent,
+				}
+				layerInfoMap[layer.Parent].ChildID = append(layerInfoMap[layer.Parent].ChildID, layer.ID)
+			} else {
+				existingLayer.ChildID = append(existingLayer.ChildID, layer.ID)
+			}
+		} else {
+			existingLayer.ParentID = layer.ID
+			existingLayer.Size = layer.UncompressedSize
+		}
+	}
+
+	// Add the Repo Tags to Top layer of each image.
+	for _, img := range imgs {
+		e, ok := layerInfoMap[img.TopLayer]
+		if !ok {
+			return nil, fmt.Errorf("image top-layer not found in LayerList! : %s ", img.TopLayer)
+		}
+		e.RepoTags = append(e.RepoTags, img.Names...)
+	}
+	return layerInfoMap, nil
+}
+
+// Stores all parent layers using which image is built in imageInfo
+func buildImageDependencyMap(i *Image, imageInfo *InfoImage, layerMap map[string]*LayerInfo, layerID string) error {
+	if layerID == "" {
+		return nil
+	}
+	ll, ok := layerMap[layerID]
+	if !ok {
+		return fmt.Errorf("lookup error: layerid  %s, not found", layerID)
+	}
+	if err := buildImageDependencyMap(i, imageInfo, layerMap, ll.ParentID); err != nil {
+		return err
+	}
+
+	imageInfo.Layers = append(imageInfo.Layers, *ll)
+	imageInfo.Size = imageInfo.Size + ll.Size
+	return nil
+}
+
+// Stores all children layers which are created using given Image.
+func buildImageDependentsMap(i *Image, imageInfo *InfoImage, layerMap map[string]*LayerInfo, layerID string, prefixSpace int, buffer *bytes.Buffer) error {
+	if layerID == "" {
+		return nil
+	}
+	ll, ok := layerMap[layerID]
+	if !ok {
+		return fmt.Errorf("lookup error: layerid  %s, not found", layerID)
+	}
+	imageInfo.Layers = append(imageInfo.Layers, *ll)
+	imageInfo.Size = imageInfo.Size + ll.Size
+	for p := 0; p < prefixSpace; p++ {
+		buffer.WriteString(" ")
+	}
+	buffer.WriteString(fmt.Sprintf("└─ ID: %s Size: %v Tags: %s\n", ll.ID[:12], units.HumanSizeWithPrecision(float64(ll.Size), 4), ll.RepoTags))
+	if len(ll.RepoTags) > 0 {
+		prefixSpace++
```

Comment:
AFAICS this means that the depths of the tree do not accurately reflect the parent/child relationships: a parent layer with no tag would be displayed at the same level as its children.

(Indenting only on a tag might be a reasonable way to format the non parent layer list, but it it seems to be a bad fit here, and it would probably be convenient if both were consistent.)

Reply:
A layer has only one parent. We do not have history or any other metadata information based on that we can classify if, particular layer is created as part of different image, except `TopLayer` of Image have some `tag` associated with it. 

In case a `tag` is removed from base image as part of image deletion, the layers becomes part of child layer and hence no child-parent hierarchy between layers.

Reply:
> We do not have history or any other metadata information based on that we can classify if, particular layer is created as part of different image

We have built the list of children; we can _precisely_ determine whether a layer is a parent, within the context of the layers that we plan to print.

> In case a `tag` is removed from base image as part of image deletion, the layers becomes part of child layer and hence no child-parent hierarchy between layers.

“layer becomes a part of child layer” does not make any sense at all.

Consider the actual layout (where indent means “is a child of the outdented parent”)
```
└─ unnamed-sharedparent
  └─ unnamed-child1
    └─ tagged-child1
  └─ unnamed-child2
    └─ tagged-child2
```

This code will AFAICS format it as
```
└─ unnamed-sharedparent
└─ unnamed-child1
└─ tagged-child1
└─ unnamed-child2
└─ tagged-child2
```
which does not show the structure at all.

---

Or, consider the actual layout (where indent means “is a child of the outdented parent”)
```
└─ unnamed-sharedparent
  └─ tagged-child1
  └─ tagged-child2
```

this code will format it as
```
└─ unnamed-sharedparent
└─ tagged-child1
  └─ tagged-child2
```
which is actively wrong.

Reply:
Looking again, my second example seems to be incorrect (or is incorrect with the current version of the code, because indentation only happens at the parent/child boundary, not between peers — so the second example would also output a linear non-indented sequence.

Either way, this indentation design seems misleading to me.

Reply:
I am sorry, but I am not able to understand this.

Lets discuss on this example, it will be easy understand exact output
```
$ sudo podman image tree docker.io/library/php:7.2-apache --whatrequires
Image ID        : e2834342b873
Tags            : [docker.io/library/php:7.2-apache]
Size            : 386.2MB
Image Layers :
└─ ID: 80715f9e8880 Size: 4.608kB Tags: [docker.io/library/php:7.2-apache]
 └─ ID: c93cbcd6437e Size: 3.573MB Tags: []
 └─ ID: dece674f3cd1 Size: 4.608kB Tags: []
 └─ ID: 834f4497afda Size: 7.168kB Tags: []
 └─ ID: bfe2ce1263f8 Size: 40.06MB Tags: []
 └─ ID: 748e99b214cf Size: 11.78kB Tags: [docker.io/library/wordpress:latest]
```
_TAG is assigned to only top layer of image_, so any layers below will go as part of child image.
also the output is from `root-layer->...->last-layer(top-layer)`

-----------
For second example, below is one example to show child parent relationship and formatting.
- `foobar1` & `foobar2` are peer & child of `docker.io/library/alpine`
- foobar1-2 is child of `foorbar1`
```
$ sudo podman image tree docker.io/library/alpine --whatrequires     
Image ID        : 196d12cf6ab1
Tags            : [docker.io/library/alpine:latest]
Size            : 4.674MB
Image Layers :
└─ ID: df64d3292fd6 Size: 4.672MB Tags: [docker.io/library/alpine:latest]
 └─ ID: acd12e3fec60 Size: 3.584kB Tags: [localhost/foobar1:latest]
  └─ ID: a72c6b818bfc Size: 2.56kB Tags: [localhost/foobar1-2:latest]
 └─ ID: 570731fd1e5d Size: 3.584kB Tags: [localhost/foobar2:latest]

```

Reply:
Both of these examples only _happen to_ work because the names are present like that.  In the first example, `podman tag 834f4497afda some-tag` will change the indentation.

In the second example, `podman remove localhost/foobar1:latest` will, again, change the indentation.